instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * <ul>
+ *   <li>Create an {@link Instrumenter} using {@link InstrumenterBuilder}. Use the builder to
+ *       configure any library-specific customizations, and also expose useful knobs to your user.
+ *   <li>Call {@link Instrumenter#start(Context, Object)} at the beginning of a request.
+ *   <li>Call {@link Instrumenter#end(Context, Object, Object, Throwable)} at the end of a request.
+ * </ul>
+ */
+public class Instrumenter<REQUEST, RESPONSE> {
+
+  /** Returns a new {@link InstrumenterBuilder}. */
+  public static <REQUEST, RESPONSE> InstrumenterBuilder<REQUEST, RESPONSE> newBuilder(
+      OpenTelemetry openTelemetry,
+      String instrumentationName,
+      SpanNameExtractor<? super REQUEST> spanNameExtractor) {
+    return new InstrumenterBuilder<>(openTelemetry, instrumentationName, spanNameExtractor);
+  }
+
+  private final Tracer tracer;
+  private final SpanNameExtractor<? super REQUEST> spanNameExtractor;
+  private final SpanKindExtractor<? super REQUEST> spanKindExtractor;
+  private final StatusExtractor<? super REQUEST, ? super RESPONSE> statusExtractor;
+  private final List<? extends AttributesExtractor<? super REQUEST, ? super RESPONSE>> extractors;
+  private final ErrorCauseExtractor errorCauseExtractor;
+
+  Instrumenter(
+      Tracer tracer,
+      SpanNameExtractor<? super REQUEST> spanNameExtractor,
+      SpanKindExtractor<? super REQUEST> spanKindExtractor,
+      StatusExtractor<? super REQUEST, ? super RESPONSE> statusExtractor,
+      List<? extends AttributesExtractor<? super REQUEST, ? super RESPONSE>> extractors,
+      ErrorCauseExtractor errorCauseExtractor) {
+    this.tracer = tracer;
+    this.spanNameExtractor = spanNameExtractor;
+    this.spanKindExtractor = spanKindExtractor;
+    this.statusExtractor = statusExtractor;
+    this.extractors = extractors;
+    this.errorCauseExtractor = errorCauseExtractor;
+  }
+
+  /**
+   * Starts a new operation to be instrumented. The {@code parentContext} is the parent of the
+   * resulting instrumented operation and should usually be {@code Context.current()}. The {@code
+   * request} is the request object of this operation. The returned {@link Context} should be
+   * propagated along with the operation and passed to {@link #end(Context, Object, Object,
+   * Throwable)} when it is finished.
+   */
+  public Context start(Context parentContext, REQUEST request) {
+    SpanKind spanKind = spanKindExtractor.extract(request);
+    SpanBuilder spanBuilder =
+        tracer
+            .spanBuilder(spanNameExtractor.extract(request))
+            .setSpanKind(spanKind)
+            .setParent(parentContext);
+
+    AttributesBuilder attributes = Attributes.builder();
+    for (AttributesExtractor<? super REQUEST, ? super RESPONSE> extractor : extractors) {
+      extractor.onStart(attributes, request);
+    }
+    attributes.build().forEach((key, value) -> spanBuilder.setAttribute((AttributeKey) key, value));
```

Comment:
why not directly calling add to the builder?

Reply:
Since it seems extremely likely that we'll use `Attributes` for metrics too (first version I'd just extract the string key/values when adding our first metric here). If performance is a concern, I'm planning on creating a special implementation of `AttributesBuilder` which is faster but unsafe for general consumption.